lib/fizzy/execute.cpp

```diff
@@ -232,43 +254,36 @@ Instance instantiate(const Module& module, std::vector<ImportedFunction> importe
     // NOTE: fill it with zeroes
     bytes memory(memory_min * PageSize, 0);
 
-    // Fill out memory based on data segments
-    for (const auto& data : module.datasec)
-    {
-        uint64_t offset;
-        if (data.offset.kind == ConstantExpression::Kind::Constant)
-            offset = data.offset.value.constant;
-        else
-            throw std::runtime_error("data initialization by imported global is not supported yet");
-
-        // NOTE: these instructions can overlap
-        assert((offset + data.init.size()) <= (memory_max * PageSize));
-        std::memcpy(memory.data() + offset, data.init.data(), data.init.size());
-    }
-
     std::vector<TypeIdx> imported_function_types =
         match_imports(module, imported_functions, imported_globals);
-
     std::vector<uint64_t> globals;
     globals.reserve(module.globalsec.size());
     // init regular globals
     for (auto const& global : module.globalsec)
     {
-        if (global.expression.kind == ConstantExpression::Kind::Constant)
-            globals.emplace_back(global.expression.value.constant);
-        else
+        // Wasm spec section 3.3.7 constrains initialization by another global to const imports only
+        // https://webassembly.github.io/spec/core/valid/instructions.html#expressions
+        if (global.expression.kind == ConstantExpression::Kind::GlobalGet &&
+            global.expression.value.global_index >= imported_globals.size())
         {
-            // initialize by imported global
-            const auto global_idx = global.expression.value.global_index;
-            // Wasm spec section 3.3.7 constrains initialization by another global to const imports
-            // only https://webassembly.github.io/spec/core/valid/instructions.html#expressions
-            if (global_idx >= imported_globals.size() || imported_globals[global_idx].is_mutable)
-            {
-                throw std::runtime_error(
-                    "global can be initialized by another global only if it's const and imported");
-            }
-            globals.emplace_back(*imported_globals[global_idx].value);
+            throw std::runtime_error(
+                "Global can be initialized by another const global only if it's imported.");
         }
+
+        const auto value = eval_constant_expression(
+            global.expression, imported_globals, module.globalsec, globals);
+        globals.emplace_back(value);
+    }
+
+    // Fill out memory based on data segments
```

Comment:
Can you create a PR (or commit) which reorders the file (according to the instantiation spec)?

Could also move the `load_from_memory `helpers above `instantiate()`.

Reply:
Did the first reorder in the first commit here.

Reply:
Did the second reorder in https://github.com/wasmx/fizzy/pull/104